Name: **Rachel Jasmine Canaman** <br>
Section: **DS4A**

<h2 align="center"><b>Laboratory Task 2</b></h2>

<div style="text-align: center;">
    <img src="image1.png" alt="A centered image" style="display: block; margin: 0 auto;">
</div>

**Objective:**  
Perform a single forward pass through a simple artificial neural network (ANN) and compute the error.

We are given:

- Input vector: \( x = \begin{bmatrix}1 \\ 0 \\ 1\end{bmatrix} \), and target output \( y = 1 \)

- Hidden unit weights: \( W_{\text{hidden}} = \begin{bmatrix} 0.2 & -0.3 \\ 0.4 & 0.1 \\ -0.5 & 0.2 \end{bmatrix} \)

- Output weights: \( w_{21} = -0.3 \), \( w_{22} = -0.2 \)

- Biases: \( \theta = \begin{bmatrix} -0.4 \\ 0.2 \\ 0.1 \end{bmatrix} \)

- Activation function (ReLU): \( f(z) = \max(0, z) \)


**Step 1: Import libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt

**Step 2: Define Inputs, Weights, and Biases**

In [2]:
# Input and target
x = np.array([1, 0, 1])   # input vector
y = 1                     # target output

# Hidden layer weights (3x2)
W_hidden = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2]
])

# Biases
theta1, theta2, theta3 = -0.4, 0.2, 0.1

# Output weights
w_out = np.array([-0.3, -0.2])


**Step 3: Forward Pass – Hidden Layer**

We compute:  
$z_{\text{hidden}} = xW_{\text{hidden}} + \theta$ <br>
$h = f(z_{\text{hidden}})$.



In [3]:
# Hidden layer pre-activation
z_hidden = x @ W_hidden + np.array([theta1, theta2])

# Apply ReLU activation
h = np.maximum(0, z_hidden)

print("Hidden pre-activations:", z_hidden)
print("Hidden activations (after ReLU):", h)

Hidden pre-activations: [-0.7  0.1]
Hidden activations (after ReLU): [0.  0.1]


**Step 4: Forward Pass – Output Layer**

We compute:  
$z_{\text{out}} = h \cdot w_{\text{out}} + \theta_3$ <br>
$\hat{y} = f(z_{\text{out}})$.

In [4]:
# Output pre-activation
z_out = h @ w_out + theta3

# Apply ReLU (for output unit)
y_hat = max(0, z_out)

print("Output pre-activation:", z_out)
print("Predicted output:", y_hat)

Output pre-activation: 0.08
Predicted output: 0.08


**Step 5: Error Computation**

We use Mean Squared Error (MSE with 1/2 factor): $E = \frac{1}{2}(y - \hat{y})^2$.

In [5]:
# Error calculation
error = 0.5 * (y - y_hat)**2
print("Error:", error)

Error: 0.4232


**Conclusion**

- The input vector \([1, 0, 1]\) was passed through the network.
- Hidden activations were:
  - $h_1 = 0$
  - $h_2 = 0.1$
- Output prediction was $\hat{y} = 0.08$.
- The error was computed as: $E = 0.4232$

This simple forward pass demonstrates how weights, biases, and activation functions combine to produce an output. It also highlights why training (backpropagation) is essential to reduce error and improve predictions.